datadog-ktor/src/commonMain/kotlin/KtorLogger.kt

```diff
@@ -0,0 +1,128 @@
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.buildJsonObject
+import kotlinx.serialization.json.put
+
+internal class KtorLogger(
+    private val scope: CoroutineScope,
+    private val logSubmitter: LogSubmission,
+    private val config: LogConfiguration,
+    private val clock: Clock = Clock.System,
+) : RestLogger {
+    private val globalAttributes = MutableStateFlow(mutableMapOf<String, Any?>())
+    private val keyedTags = MutableStateFlow(mutableMapOf<String, String>())
+    private val unkeyedTags = MutableStateFlow(mutableListOf<String>())
+
+    override fun log(level: Logger.Level, message: String, attributes: Map<String, Any?>?, throwable: Throwable?) {
+        val tagString = keyedTags.value.entries
+            .map { (key, value) ->
+                "$key:$value"
+            }.plus(unkeyedTags)
+            .joinToString(separator = ",")
+
+        val log = buildJsonObject {
+            // Add global attributes
+            globalAttributes.value.forEach { entry -> put(entry.key, wrapInJsonPrimitive(entry.value)) }
+            // Add event-specific attributes
+            attributes?.forEach { entry -> put(entry.key, wrapInJsonPrimitive(entry.value)) }
+            // Add reserved attributes
+            put("message", message)
+            if (config.source != null) put("ddsource", config.source)
+            if (tagString.isNotEmpty()) put("ddtags", tagString)
+            if (config.host != null) put("hostname", config.host)
+            if (config.service != null) put("service", config.service)
+            put("status", level.name.lowercase())
+            put("timestamp", clock.now().toEpochMilliseconds())
+        }
+        scope.launch { logSubmitter.submitLogs(listOf(log)) }
+    }
+
+    override fun debug(message: String, attributes: Map<String, Any?>?, throwable: Throwable?) {
+        log(Logger.Level.Debug, message, attributes, throwable)
+    }
+
+    override fun info(message: String, attributes: Map<String, Any?>?, throwable: Throwable?) {
+        log(Logger.Level.Info, message, attributes, throwable)
+    }
+
+    override fun warn(message: String, attributes: Map<String, Any?>?, throwable: Throwable?) {
+        log(Logger.Level.Warn, message, attributes, throwable)
+    }
+
+    override fun error(message: String, attributes: Map<String, Any?>?, throwable: Throwable?) {
+        log(Logger.Level.Error, message, attributes, throwable)
+    }
+
+    override fun assert(message: String, attributes: Map<String, Any?>?, throwable: Throwable?) {
+        log(Logger.Level.Assert, message, attributes, throwable)
+    }
+
+    override fun verbose(message: String, attributes: Map<String, Any?>?, throwable: Throwable?) {
+        log(Logger.Level.Verbose, message, attributes, throwable)
+    }
+
+    override fun addTag(tag: String) {
+        unkeyedTags.update { tags ->
+            tags.add(tag)
+            return@update tags
+        }
+    }
```

Comment:
If you define `unkeyedTags` as:

```kotlin
private val unkeyedTags = MutableStateFlow(emptyList<String>())
```

```suggestion
    override fun addTag(tag: String) = unkeyedTags.update { it + tag }
```